product_mandatory_products/controller/main.py

```diff
@@ -0,0 +1,149 @@
+            and "kwargs" in kw
+            and not kw["kwargs"]["context"].get("display_optional")
+        ):
+            product = request.env["product.product"].browse(int(product_id))
+            combination = request.env["product.template.attribute.value"].browse(
+                variant_values
+            )
+            add_qty = int(kw.get("add_qty", 1))
+
+            no_variant_attribute_values = combination.filtered(
+                lambda product_template_attribute_value: product_template_attribute_value.attribute_id.create_variant  # noqa : E501
+                == "no_variant"
+            )
+            if no_variant_attribute_values:
+                product = product.with_context(
+                    no_variant_attribute_values=no_variant_attribute_values
+                )
+            return request.env["ir.ui.view"]._render_template(
+                "product_mandatory_products.mandatory_products_modal",
+                {
+                    "product": product,
+                    "combination": combination,
+                    "add_qty": add_qty,
+                    "parent_name": product.name,
+                    "variant_values": variant_values,
+                    "pricelist": pricelist,
+                    "handle_stock": handle_stock,
+                    "already_configured": kw.get("already_configured", False),
+                },
+            )
+        return super(ProductConfigurator, self)._show_advanced_configurator(
+            product_id, variant_values, pricelist, handle_stock, **kw
+        )
+
+
+class WebsiteSaleExtend(WebsiteSale):
+    KW = []
+
+    def update_product_and_options(self, product_and_options):
+        # Set new parent_unique_id
+        uid = json.loads(product_and_options)[0]["unique_id"]
+        for i in self.KW:
+            i["parent_unique_id"] = uid
+        # Update old root product with new values
+        self.KW[0].update(json.loads(product_and_options)[0])
+        # Add new variants in list
+        self.KW.extend(json.loads(product_and_options)[1:])
+        return json.dumps(self.KW)
+
+    @http.route(
+        ["/shop/cart/update_option"],
+        type="http",
+        auth="public",
+        methods=["POST"],
+        website=True,
+        multilang=False,
+    )
+    def cart_options_update_json(
+        self, product_and_options, goto_shop=None, lang=None, **kwargs
+    ):
+        if goto_shop == "true":
+            if self.KW:
+                # Call only when there is mandatory products.
+                product_and_options = self.update_product_and_options(
+                    product_and_options
+                )
+            res = super(WebsiteSaleExtend, self).cart_options_update_json(
+                product_and_options, goto_shop=goto_shop, lang=lang, **kwargs
+            )
+            self.KW = []
+            return res
+        else:
+            self.KW.clear()
+            self.KW.extend(json.loads(product_and_options))
```

Comment:
I have some question about this....is the content of KW subject to "surviving" different route calls?